### PR TITLE
Implement golangci linters

### PR DESCRIPTION
**Pull request by RMPall**
### What

- This adds the golangci linter to the dp-search-reindex-api repo. This PR enables the following linters via .golangci.yml
`
    - deadcode
    - depguard
    - dogsled
    - errcheck
    - gochecknoinits
    - goconst
    - gocritic
    - gocyclo
    - gofmt
    - goimports
    - revive
    - gosec
    - gosimple
    - govet
    - ineffassign
    - nakedret
    - staticcheck
    - structcheck
    - stylecheck
    - typecheck
    - unconvert
    - unused
    - varcheck
    - whitespace
    - gocognit
    - prealloc`

This PR also attempts to fix the existing lint issues with the repo. **Please Note** in .golangci.yaml linter configuration file there is an parameter under issues attribute reffered to as `new` I have set it to false (default), but when you set it to true then it will ignore all the existing lint issues and will flag only the new lint issues. 

### How to review

Please have a look into the .golangcilint.yml file and see if the linters listed here make sense , also if you would like to add any additional linters we can always do that . For the first pass I think we have enough linters enabled (mentioned above). 

### Who can review

Anyone except me

### CODE REVIEW

**Review by cookel2 — COMMENTED** [review]
new lines needed at ends of some files

File: project_generation/content/templates/event/.golangci.yml.tmpl [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        - dupl
+        - gosec
+  new: false
+
+# golangci.com configuration
+# https://github.com/golangci/golangci/wiki/Configuration
+service:
+  golangci-lint-version: 1.43.x # use the fixed version to not introduce new linters unexpectedly
+  prepare:
+    - echo "here I can run custom commands, but no preparation needed for this repo"
```
[the remark this answers is not in the document]
Comment (rahulPalliyalil) [reply project_generation/content/templates/event/.golangci.yml.tmpl]: Done in update1.

File: project_generation/content/templates/base-app/.golangci.yml.tmpl [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        - dupl
+        - gosec
+  new: false
+
+# golangci.com configuration
+# https://github.com/golangci/golangci/wiki/Configuration
+service:
+  golangci-lint-version: 1.43.x # use the fixed version to not introduce new linters unexpectedly
+  prepare:
+    - echo "here I can run custom commands, but no preparation needed for this repo"
```
[the remark this answers is not in the document]
Comment (rahulPalliyalil) [reply project_generation/content/templates/base-app/.golangci.yml.tmpl]: Done in update 1.

**Comment by cookel2** [comment]
Looks good @rahulPalliyalil

File: project_generation/content/templates/api/.golangci.yml.tmpl [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  funlen:
+  # lines: 100
+  # statements: 100
+
+linters:
+  # please, do not use `enable-all`: it's deprecated and will be removed soon.
+  # inverted configuration with `enable-all` and `disable` is not scalable during updates of golangci-lint
+  disable-all: true
+  enable:
+    - deadcode
```
Comment (nshumoogum) [inline project_generation/content/templates/api/.golangci.yml.tmpl:51]: I believe `deadcode` and `varcheck` have been deprecated - worth removing and checking if there are others

File: project_generation/content/templates/base-app/Makefile.tmpl [hunk 4]
```diff
@@ -15,7 +15,8 @@ audit:
 
 .PHONY: lint
 lint:
-	exit
+	go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.43.0
```
Comment (nshumoogum) [inline project_generation/content/templates/base-app/Makefile.tmpl:18]: Worth updating to latest version

**Review by nshumoogum — CHANGES_REQUESTED** [review]

**Comment by nshumoogum** [comment]
Closing - another PR with latest changes https://github.com/ONSdigital/dp-cli/pull/132

**Pull request closed without merging**